hw/drivers/chg_ctrl/adp5061/src/adp5061.c

```diff
@@ -0,0 +1,583 @@
+    /* EN_EOC, EN_CHG */
+    .functional_settings_1 = 0x07,
+    /* VSYSTEM = 4.3V, IDEAL_DIODE = 01 */
+    .functional_settings_2 = 0x08,
+    /* Int THR enable */
+    .interrupt_enable = 0xFF,
+    /* TBAT_SHR = 30s, VBAT_SHR = 2.4V*/
+    .battery_short = 0x84,
+    /* SYS_EN_SET = 1 */
+    .iend = 0x01,
+};
+
+
+#if MYNEWT_VAL(ADP5061_INT_PIN) >= 0
+/**
+* ADP5061 interrupt hanlder CB
+* gets interrupt status and prints to console
+*/
+
+static void
+adp5061_event(struct os_event *ev)
+{
+    assert(ev);
+    struct adp5061_dev *dev = ev->ev_arg;
+
+    /* Interrupt generated by charger triggers out of schedule read */
+    charge_control_read(&dev->a_chg_ctrl,
+            CHARGE_CONTROL_TYPE_STATUS | CHARGE_CONTROL_TYPE_FAULT,
+            NULL, NULL, OS_TIMEOUT_NEVER);
+}
+
+/**
+* ADP5061 interrupt handler structure
+*/
+static struct os_event interrup_handler = {
+    .ev_cb = adp5061_event,
+};
+
+/**
+* ADP5061 IRQ
+* Add interrupt handling task to queue
+*/
+static void
+adp5061_isr(void *arg){
+    os_eventq_put(os_eventq_dflt_get(), &interrup_handler);
+}
+#endif
+
+static int
+adp5061_write_configuration(struct adp5061_dev *dev)
+{
+    int rc = 0;
+    const uint8_t regs[] = {
+            dev->a_cfg.vinx_pin_settings,
+            dev->a_cfg.termination_settings,
+            dev->a_cfg.charging_current,
+            dev->a_cfg.voltage_thresholds,
+            dev->a_cfg.timer_settings,
+            dev->a_cfg.functional_settings_1,
+            dev->a_cfg.functional_settings_2,
+            dev->a_cfg.interrupt_enable
+    };
+
+    rc = adp5061_set_reg(dev, 0x10, dev->a_cfg.battery_short);
+    if (rc) {
+        goto err;
+    }
+    rc = adp5061_set_reg(dev, 0x11, dev->a_cfg.iend);
+    if (rc) {
+        goto err;
+    }
+    rc = adp5061_set_regs(dev, 0x02, regs, sizeof(regs));
+    if (rc) {
+        goto err;
+    }
+err:
+    return rc;
+}
+
+int
+adp5061_set_config(struct adp5061_dev *dev,
+        const struct adp5061_config *cfg)
+{
+    int rc = 0;
+
+    dev->a_cfg = *cfg;
+
+    rc = adp5061_write_configuration(dev);
+
+    return rc;
+}
+
+int
+adp5061_get_reg(struct adp5061_dev *dev, uint8_t addr, uint8_t *value)
+{
+    int rc = 0;
+    uint8_t payload;
+    struct hal_i2c_master_data data_struct = {
+        .address = dev->a_chg_ctrl.cc_itf.cci_addr,
+        .len = 1,
+        .buffer = &payload
+    };
+    /* Register write */
+    payload = addr;
+    rc = hal_i2c_master_write(dev->a_chg_ctrl.cc_itf.cci_num, &data_struct,
+            OS_TICKS_PER_SEC / 10, 1);
+    if (rc) {
+        goto err;
+    }
+    /* Read one byte back */
+    payload = addr;
+    rc = hal_i2c_master_read(dev->a_chg_ctrl.cc_itf.cci_num, &data_struct,
+            OS_TICKS_PER_SEC / 10, 1);
+    *value = payload;
+
+err:
+    return rc;
+}
+
+int
+adp5061_set_reg(struct adp5061_dev *dev, uint8_t addr, uint8_t value)
+{
+    int rc = 0;
+    uint8_t payload[2] = {};
+    payload[0] = addr;
+    payload[1] = value;
```

Comment:
We try to keep variable declarations in the top, and not declare vars in the middle. To make that true here, you could do e.g.:
uint8_t payload[2] = { addr, value };